wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,709 @@
+# cuML Python Estimators Developer Guide
+
+This guide is meant to help developers follow the correct patterns when creating/modifying any class that inherits from `cuml.common.base.Base`.
+
+Start by reading the following:
+
+1. [SkLearn Estimator Docs](https://scikit-learn.org/stable/developers/develop.html)
+   1. `cuml` Estimator design follows Sklearn very closely. We will only cover portions where our design differs from this document
+   2. Pay close attention to these sections as these topics have caused pain points in the past:
+      1. [Instantiation](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+      2. [Estimated Attributes](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+      3. [`get_params` and `set_params`](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+      4. [cloning](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+2. [SkLearn Docstring Guide](https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation)
+   1. We follow the same guidlines for speficying array-like objects, array shapes, dtypes, and default values
+
+## Background
+
+Some background is necessary to understand the design of estimators and how to work around any non-standard situations.
+
+### Input and Output Types in cuML
+
+In `cuml` we support both ingesting and generating a variety of different n-dimensional array types. Estimators should be able to accept and return any array type. The types that are supported as of release 0.17:
+
+ - cuDF DataFrame or Series
+ - Pandas DataFrame or Series
+ - NumPy Arrays
+ - Numba Device Arrays
+ - CuPy arrays
+ - CumlArray type (Internal to the `cuml` API only.)
+
+When converting between types, it's important to minimize the CPU<->GPU type conversions as much as possible. Conversions such as NumPy -> CuPy or Numba -> Pandas DataFrame will incur a performance penalty as memory is copied from device to host or viceversa.
+
+Converting between types of the same device, i.e. CPU<->CPU or GPU<->, do not have as significant of a penalty.
+
+Finally, conversions between Numba<->CuPy<->CumlArray incur the least amount of overhead since only the device pointer is moved from one class to another.
+
+Internally, all array's should be converted to `CumlArray` as much as possible since it is compatible with all output types and can be easily converted.
+
+### Speficying the Array Output Type
+
+Users can choose which array type should be returned by cuml by either: 
+1. Individually setting the output_type property on an estimator class (i.e `Base(output_type="numpy")`)
+2. Globally setting the `cuml.global_output_type`
+
+Note: Setting the global output type will take precedence over any value in `Base.output_type`
```

Comment:
I hear what you are saying. I went back and forth on this a few times. Ultimately I went with a design that allows for maximum flexibility while respecting existing functionality, without making the design (and therefore the code) too complex. I found that using `Base.output_type` to override the `cuml.global_output_type` was pretty restrictive. For example, assuming `Base.output_type` overrides `cuml.global_output_type`:

- You can't temporarily change multiple estimators or one-off algorithms at the same time.
   - If you wanted to change the output_type temporarily (for example, to send to an sklearn algorithm) you would have to update the `output_type` of all estimators, call some function, then revert the `output_type` setting.
- `cuml.global_output_type` becomes the "default" output type, not the "global" setting.
   - This is because estimators would only care about `cuml.global_output_type` in the constructor, and only then if `output_type` wasnt specified as an argument.
   - This design might make more sense if estimator's default value could store a reference to `cuml.global_output_type` instead of copying the value during construction.
- The utility of `cuml.using_output_type()` is significantly decreased.
   - Changing the `cuml.global_output_type` temporarily with a context manager would only be useful during estimator construction
- We would need different methods to change the `output_type` internally vs. externally.
   - Currently, `cuml.using_output_type()` can be used in cython to cast a CumlArrayDescriptor if a developer needs a specific type of array. If `cuml.using_output_type()` could no longer accomplish this task, we would need a different method that works similar to the current design.

Ultimately, I prefer having `cuml.global_output_type` override `Base.output_type`. It allows you to do things like:
```python
estA = MyEstimator().fit(some_cupy_data)
estB = MyOtherEstimator().fit(some_cupy_data)

cuml.using_output_type("numpy"):
   sk_r2_score = sklearn.r2_score(estA.my_cupy_array_, estB.my_cupy_array_)

```